### PR TITLE
Updating deprecated CrossProduct in the StructuralMechanicsApplication

### PR DESCRIPTION
**Pull request by loumalouomega**
Additioanlly the Doxygen from math_utils has been updated (to avoid errors in the Doxygen documentation generation)

### CODE REVIEW

**Comment by loumalouomega** [comment]
Don't merge, the thin shell tests are failing, I did some mistake my bad

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: just to tell that it would automatically detect the template parameters. you do not need to put them explicitly (it is correct if you do of course)

File: applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp:1989]: IMPORTANT: not all the instances of CrossProduct have to be changed, only the ones for which you get a deprecation warning

**Review by RiccardoRossi — COMMENTED** [review]
just one comment: not all the cases of 

a = CrossProduct(b,c) 

are deprecated. Only the ones for which the order was changed

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: https://stackoverflow.com/questions/6359592/c-auto-detection-of-template-arguments You cannot...

File: applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp:1989]: I know, I added a const

**Comment by loumalouomega** [comment]
@RiccardoRossi I added default values to the templates, by default it is an array_1d<double, 3> for the first type, and the second and third type are same type than the first type. This reduces the number of template arguments to one in the normal cases, and to zero for array_1d<double, 3>. I chose array_1d<double, 3> because that way I can remove a "duplicated" method of CrossProduct

File: kratos/utilities/math_utils.h
Comment (RiccardoRossi) [inline kratos/utilities/math_utils.h]: why did u remove it?

File: kratos/utilities/math_utils.h
Comment (RiccardoRossi) [inline kratos/utilities/math_utils.h]: why?

you could already wrote simply

UnitCrossProduct(a,b,c)

now you can only do if it matches your default

File: kratos/utilities/math_utils.h
Comment (RiccardoRossi) [inline kratos/utilities/math_utils.h]: same

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
i don t agree with your latest commit. i think you just made it worst by adding the defaults and taking out the specialization

File: kratos/utilities/math_utils.h
Comment (RiccardoRossi) [reply kratos/utilities/math_utils.h]: just to add that if for any reason the 

       UnitCrossProduct(a,b,c)

does not work, just try

      UnitCrossProduct<>(a,b,c)

File: kratos/utilities/math_utils.h
Comment (RiccardoRossi) [reply kratos/utilities/math_utils.h]: just to clarify, it was intended to provide an explicit template instantiation of the  general template before, although most probably it misses the "template<>" in front of it to achieve the objective

File: kratos/utilities/math_utils.h
Comment (RiccardoRossi) [reply kratos/utilities/math_utils.h]: for the reference on why it works, take a look at 

http://en.cppreference.com/w/cpp/language/template_argument_deduction

Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: Done

**Comment by loumalouomega** [comment]
Ping @RiccardoRossi

File: applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp:2004]: what do you mean? if you get the deprecation warning you need to change this. Also as a side remark, multiplying by -1 is quite inefficient in this context (you could use -r_13)

Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp:2004]: i mean, i would change this as

     Vector vec1(3), vec2(3);
     MathUtils<double>::CrossProduct(vec1,e_24,-r_13 );
     MathUtils<double>::CrossProduct(vec2,e_24,r_13);

however this code is sort of stupid. since vec2 = -vec1...

File: kratos/utilities/math_utils.h
Comment (RiccardoRossi) [inline kratos/utilities/math_utils.h]: ok, but also check the size of c

File: kratos/utilities/math_utils.h
Comment (RiccardoRossi) [inline kratos/utilities/math_utils.h:936]: make the same size check as for the CrossProduct if you are in idebug

File: applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp:2004]: @RiccardoRossi , I already tried that, it doesn't work. This code is not mine, feel free of correct it and check like it doesn't behave as expected, so probably there is some kind of memory error

File: kratos/utilities/math_utils.h
Comment (loumalouomega) [reply kratos/utilities/math_utils.h:936]: No, because this calls the CrossProduct, so it is done there

File: kratos/utilities/math_utils.h
Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: The problem is that c can be a Vector of zero size, then we need to resize, but can be an array_1d of size 2 for example, with no possibility of resize

File: kratos/utilities/math_utils.h
Comment (RiccardoRossi) [reply kratos/utilities/math_utils.h:936]: you are right

File: applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp:2004]: it will work now because of the templating we introduced (an expression will be passed in this case)

File: kratos/utilities/math_utils.h
Comment (RiccardoRossi) [reply kratos/utilities/math_utils.h]: correct, so in debug throw an error if the size of c is not correct

Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: I added but some problems didn't work, that means I should correct all the vectors not initialized?

**Comment by loumalouomega** [comment]
PING

**Comment by loumalouomega** [comment]
PING

File: kratos/utilities/math_utils.h
Comment (RiccardoRossi) [reply kratos/utilities/math_utils.h]: i think you do need to have this check. If it throws an error it means it is wrong... also see my comments in #1100 about verifying if aliasing exists

**Review by RiccardoRossi — APPROVED** [review]
i am approving provided that the additional tests are added in the CrossProduct

File: kratos/utilities/math_utils.h
Comment (RiccardoRossi) [inline kratos/utilities/math_utils.h]: i think you must make a cast, for example for std::size_t for the check to work. for example

KRATOS_ERROR_IF((std::size_t)&c == (std::size_t)&a)

(or even better with a static cast of the pointer direction to std::size_t

Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: Yes, I had problems because of the type, I was going to ask what to do

**Comment by loumalouomega** [comment]
@RiccardoRossi Finally I added an additional method using std::if_enable for the alias (I tried wioth pointers, refrences, etc...,etc...), this was the only option that it worked

**Comment by loumalouomega** [comment]
#1100

**Pull request merged**
